One slice of PR3153 - primarily Statements (#3423)
dafny-lang/dafny: docs/DafnyRef/Statements.md

Examples:
<!-- %check-resolve -->
```dafny
class A { var f: int }
method m(a: A) {
  label x: while true {
       if (*) { break x; }
  }
  a.f := 0;
  label y:
  a.f := 1;
  assert old@y(a.f) == 1;
}
```

A labeled statement is just 
- the keyword `label` 
- followed by an identifier, which is the label, 
- followed by a colon 
- and a statement. 

The label may be
referenced in a `break` or `continue` statement within the labeled statement
(see [Section 20.2](#sec-break-continue-statement)). That is, the break or continue that
mentions the label must be _enclosed_ in the labeled statement.

The label may also be used in an `old` expression ([Section 21.22](#sec-old-expression)). In this case, the label
must have been encountered during the control flow en route to the `old`
expression. We say in this case that the (program point of the) label _dominates_
the (program point of the) use of the label.
Similarly, labels are used to indicates previous states in calls of [two-state predicates](#sec-two-state),
[fresh](#sec-fresh-expression) expressions, [unchanged expressions](#sec-unchanged-expression), 

> and [allocated](#sec-allocated-expression) expressions.

Author: Thanks

evaluated, then they are assigned to the out-parameters, and then the
method terminates.

## 20.5. Yield Statement ([grammar](#g-yield-statement)) {#sec-yield-statement}

A yield statement can only be used in an iterator.

> See [Section 15](#sec-iterator-types) for more details

Reviewer: ```suggestion
See [iterator types](#sec-iterator-types) for more details
```
Depending on a section number is irrelevant if the document has clickable hyperrefs.

Author: But for those who read a paper copy they are helpful!

Reviewer: Facts:
* I think 99.99% of people will use a clickable version (the fact is that I believe this)
* Manually writing section numbers is our maintenance burden that even LaTeX users don't have.
* For those to dare to print this reference guide on not less than 190 sheets of paper which is roughly two pounds of paper if you print two-sided, they will have the problem that it's always outdated. And if they print less, then these links don't serve any purpose.

assigned to the yield parameters, and then the iterator
yields.

## 20.6. Update and Call Statements ([grammar](#g-update-and-call-statement)) {#sec-update-and-call-statement}

Examples:
<!-- %check-resolve -->
```dafny

> class C { var f: int }

Reviewer: class D {
  var i: int
  constructor(i: int) {
    this.i := i;
  }
}

method q(i: int, j: int) {}
method r() returns (s: int, t: int) { return 2,3; }
method m() {

>   var ss: int, tt: int, c: C?, a: array<int>;

Reviewer: ```suggestion
  var ss: int, tt: int, c: C?, a: array<int>, d: D?;
```

Author: What does this add, given there is `c: C?`

Author: OK

  q(0,1);
  ss, c.f := r();

>   c := new C;

Reviewer: ```suggestion
  c := new C;
  d := new d(2);
```

Author: OK

To use this form of update,

 * if the RHS of the update-with-failure statement is a method call, the first out-parameter of the callee must be failure-compatible
 * if instead, the RHS of the update-with-failure statement is one or more expressions, the first of these expressions must be a value with a failure-compatible type
 * the caller must have a first out-parameter whose type matches the output of `PropagateFailure` applied to the first output of the callee, unless an
`expect`, `assume`, or `assert` keyword is used after `:-` (cf. [Section 20.7.7](#sec-failure-return-keyword)).
 * if the failure-compatible type of the RHS does not have an `Extract` member,
then the LHS of the `:-` statement has one less expression than the RHS

> (or than the number of out-parameters from the method call), the value of the first out-parameter or expression being dropped

Reviewer: Hem, how is the value of the first out-parameter is being dropped? I though it was taken and just extracted as a boolean.

Author: This is the case of a non-value-carrying failure-compatible type, like Result<T>. There is no value to extract so nothing to assign.

Reviewer: Still not clear to me in this context. Can you provide an example?

Author: There are examples in the next two subsections in the text. Here is another one:
```
module {:options "--function-syntax:4"} M {
  datatype Outcome<E> = Pass | Fail(error: E)
  {
    predicate IsFailure() {
      Fail?
    }
    function PropagateFailure(): Outcome<E>
      requires Fail?
    {
      Fail(this.error)
    }
    // Note: no Extract method
  }

  method m() returns (r: Outcome<string>, t: int) {
    return Fail("bad"), 42;
  }

  method mm() returns  (r: Outcome<string>) {
    var t, tt := m();
    tt :- m();
  }
}
```

 * if the failure-compatible type of the RHS does not have an `Extract` member,
then the LHS of the `:-` statement has one less expression than the RHS
(or than the number of out-parameters from the method call), the value of the first out-parameter or expression being dropped

> (see the discussion and examples in [Section 20.7.2](#sec-simple-fc-return}

Reviewer: ```suggestion
(see the discussion and examples in [Section 20.7.2](#sec-simple-fc-return))
```
Again, I don't especially like numbered sections, but I'm ok with it for now.

There are several points to note.

* The first out-parameter of the callee is special.
  It has a special type and that type indicates that the value is inspected to see if an abrupt return
  from the caller is warranted.
  This type is often a datatype, as shown in the examples above, but it may be any type with the appropriate members.
* The restriction on the type of caller's first out-parameter is
  just that it must be possible (perhaps through generic instantiation and type inference, as in these examples) 
  for `PropagateFailure` applied to the failure-compatible output from the callee to produce a value of 
  the caller's first out-parameter type.
  If the caller's first out-parameter type is failure-compatible (which it need not be),
  then failures can be propagated up the call chain.

>   If the keyword form of the statement is used, then no `PropagateFailure` member

Reviewer: Can you explicit ":- assume" for example?

Author: Done

<!-- %no-check -->
```dafny
var x, y : int;
var x, y := 5, 6;
var x, y :- m();
var x, y :| 0 < x + y < 10;

> var (x, y) := makePair();

Reviewer: ```suggestion
var (x, y) := makePair();
var Cons(x, y) = ConsMaker();
```

unspecified boolean value is returned. The value returned
may be different each time it is executed.

## 20.10. Binding Guards ([grammar](#g-binding-guard)) {#sec-binding-guards}

Examples (in `if` statements):
<!-- %check-resolve-warn Statements.13.expect -->
```dafny
method m(i: int) {
  ghost var k: int;

>   if i, j :| 0 < i+j < 10 { k := 0; } else { k := 1; }

Reviewer: ```suggestion
  if i, j :| 0 < i+j < 10 {
    k := 0;
  } else {
    k := 1;
  }
```
Let's not artificially reduce the number of LoC changed :-)

}
```

## 20.11. If Statement ([grammar](#g-if-statement)) {#sec-if-statement}

Examples:
<!-- %check-resolve-warn Statements.14.expect -->
```dafny
method m(i: int) {
  var x: int;

>   if i > 0 { x := i; } else { x := -i; }
>   if * { x := i; } else { x := -i; }
>   if i: nat, j: nat :| i+j<10 { assert i < 10; }
>   if i == 0 { x := 0; } else if i > 0 { x := 1; } else { x := -1; }
>   if 

Reviewer: ```suggestion
  if i > 0 {
    x := i;
  } else {
    x := -i;
  }
  if * {
    x := i;
  } else {
    x := -i;
  }
  if i: nat, j: nat :| i+j<10 {
    assert i < 10;
  }
  if i == 0 {
    x := 0;
  } else if i > 0 {
    x := 1;
  } else {
    x := -1;
  }
  if 
```

  }
```

> The then alternative of the if-statement must be block statement;

Reviewer: ```suggestion
The then alternative of the if-statement must be a block statement;
```

the else alternative may be either a block statement or another if statement.
The condition of the if statement need not (but may) be enclosed in parentheses.

> An if-statement with a binding guard is a ghost statement.

Reviewer: Not true:
```dafny
method Test(i: int) returns (j: int) decreases i  {
  j := 1;
  if x :| 0 <= x < i < 10 {
    j := Test(x);
    j := j + 1;
  }
}
```
is not throwing any error. But the IDE highlights it  sometimes incorrectly as ghost

Author: OK - fixed.

> An if statement with `*` for a guard is non-deterministic.

Reviewer: ```suggestion
An if statement with `*` for a guard is non-deterministic and ghost.
```

Author: Fixed also.

at least one of the guards to evaluate to `true` (that is, `if-case`
statements must be exhaustive: the guards must cover all cases).

> In the if-with-cases, a seeqneuce of statements may follow the `=>`; it

Reviewer: ```suggestion
In the if-with-cases, a sequence of statements may follow the `=>`; it
```

need not be a block statement. Also the sequence of cases may be enclosed in 
braces but need not be.

The form that used `...` (a refinement feature) as the guard is deprecated.

## 20.12. While Statement ([grammar](#g-while-statement)) {#sec-while-statement}

Examples:
<!-- %check-resolve -->
```dafny
method m() {
  var i := 10;
  while 0 < i
    invariant 0 <= i <= 10;
    decreases i;
  {
    i := i-1;
  }
  while * {}
  i := *;
  while 
     decreases if i < 0 then -i else i
  {
     case i < 0 => i := i + 1;
     case i > 0 => i := i - 1;

>   }

Reviewer: I had looked at the grammar before but did not realized the full shape of while -case loops, very interesting !

Author: OK

  }
```

The then alternative of the if-statement must be a block statement;
the else alternative may be either a block statement or another if statement.
The condition of the if statement need not (but may) be enclosed in parentheses.

> An if statement with a binding guard is non-deterministic;

Reviewer: ```suggestion
An if statement with a binding guard is non-deterministic (unless you can prove the unicity separatedly);
```

Author: Even if uniqueness can be proved, --enforce-determinism will not allow the construct to be compiled.

verification. This suppresses attempts to check assertions (like invariants)
that would occur at the end of the loop. Eventually, however a body must
be provided; the compiler will not compile a method containing a body-less
for-loop. There is more discussion about bodyless loops in [Section 20.14.4](#sec-bodyless-constructs).

## 20.14. Loop Specifications {#sec-loop-specifications}
For some simple loops, such as those mentioned previously, Dafny can figure
out what the loop is doing without more help. However, in general the user
must provide more information in order to help Dafny prove the effect of
the loop. This information is provided by a _loop specification_. A
loop specification provides information about invariants, termination, and
what the loop modifies.
For additional tutorial information see [@KoenigLeino:MOD2011] or the
[online Dafny tutorial](../OnlineTutorial/guide).

> ### 20.14.1. Loop invariants {#sec-loop-invariants}

Reviewer: Good catch

other body-less constructs above, the verifier is silently happy with a body-less
`forall` statement, but the compiler will complain.

## 20.15. Match Statement ([grammar](#g-match-statement)) {#sec-match-statement}

Examples:
<!-- %no-check -->
```dafny

> match x case 1 => print x; case 2 => var y := x*x; print y; case _ => print "Other";
> match list { case Nil => {} case Cons(head,tail) => print head; }

Reviewer: ```suggestion

match list {
  case Nil => {}
  case Cons(head,tail) => print head;
}
match x
case 1 =>
  print x;
case 2 =>
  var y := x*x;
  print y;
case _ =>
  print "Other";
  // Any statement after is captured in this case.
```

Author: I was wanting to keep the examples compact, but we can go with the official formatting.

will complain if it finds an `assume` anywhere where it has not
been replaced through a refinement step.

Using an `{:axiom}` attribute makes the claim that the assume statement is
OK because it is known outside the Dafny program to be true.
The verifier will not complain about it, but it is the user's 

> responsibility to be absolutelu=y sure that the proposition is

Reviewer: ```suggestion
responsibility to be absolutely sure that the proposition is
```

```
Without the [`{:opaque}`](#sec-opaque) attribute, the assertion is valid; with the attribute it cannot be proved because the body if the
function is not visible. However if a `reveal f();` statement is inserted before the assertion, the proof succeeds.
Note that the pseudo-function-call in the `reveal` statement is written without arguments and serves to mark `f` as a function name
instead of a label.

## 20.21. Forall Statement ([grammar](#g-forall-statement)) {#sec-forall-statement}

Examples:
<!-- %no-check -->
```dafny

> forall i | 0 <= i < a.Length { a[i] := 0; }
> forall i | 0 <= i < 100 { P(i); } // P a lemma
> forall i | 0 <= i < 100 ensures i < 1000 {  } 

Reviewer: ```suggestion
forall i | 0 <= i < a.Length {
  a[i] := 0;
}
forall i | 0 <= i < 100 {
  P(i); // P a lemma
}
forall i | 0 <= i < 100
  ensures i < 1000 {
} 
```

```

The `forall` statement executes the body
simultaneously for all quantified values in the specified quantifier domain.

> See [Section 2.6.4](#sec-quantifier-domains) for more details on quantifier domains.

Reviewer: ```suggestion
You can find more details about [quantifier domains here](#sec-quantifier-domains).
```
